pkg/csr/csr.go

```diff
@@ -0,0 +1,248 @@
+		Type:           certificatesv1.CertificateApproved,
+		Status:         "True",
+		Message:        "This CSR was approved by the WMCO CSR Approver",
+		LastUpdateTime: metav1.Now(),
+		Reason:         "WMCOApproved",
+	})
+
+	if _, err := c.k8sClientSet.CertificatesV1().CertificateSigningRequests().UpdateApproval(context.Background(),
+		c.Name, c.CertificateSigningRequest, metav1.UpdateOptions{}); err != nil {
+		return errors.Wrapf(err, "could not update conditions for CSR approval %s", c.Name)
+	}
+	c.log.Info("CSR approved by WMCO CSR Approver", "CSR", c.Name)
+	return nil
+}
+
+// isNodeClientCert returns true if the CSR is from a  kube-apiserver-client-kubelet signer
+// reference: https://kubernetes.io/docs/reference/access-authn-authz/certificate-signing-requests/#kubernetes-signers
+func (c *Approver) isNodeClientCert(x509cr *x509.CertificateRequest) bool {
+	if !reflect.DeepEqual([]string{nodeGroup}, x509cr.Subject.Organization) {
+		return false
+	}
+	if (len(x509cr.DNSNames) > 0) || (len(x509cr.EmailAddresses) > 0) || (len(x509cr.IPAddresses) > 0) {
+		return false
+	}
+	// Check usages include: digital signature, key encipherment and client auth
+	if !c.hasUsages(kubeletClientUsages) {
+		return false
+	}
+	if !strings.HasPrefix(x509cr.Subject.CommonName, NodeUserNamePrefix) {
+		return false
+	}
+	return true
+}
+
+// hasUsages verifies if the required usages exist in the CSR spec
+func (c *Approver) hasUsages(usages []certificatesv1.KeyUsage) bool {
+	if len(usages) != len(c.Spec.Usages) {
+		return false
+	}
+
+	usageMap := map[certificatesv1.KeyUsage]struct{}{}
+	for _, u := range usages {
+		usageMap[u] = struct{}{}
+	}
+
+	for _, u := range c.Spec.Usages {
+		if _, ok := usageMap[u]; !ok {
+			return false
+		}
+	}
+
+	return true
+}
+
+// getDnsAddresses returns list of dns addresses corresponding to the given IPv4 address
+func getDnsAddresses(ipAddr string) []string {
+	addr, _ := net.LookupAddr(ipAddr)
+	return addr
+}
+
+// ParseCSR extracts the CSR from the API object and decodes it.
+func ParseCSR(csr []byte) (*x509.CertificateRequest, error) {
+	if len(csr) == 0 {
+		return nil, fmt.Errorf("CSR request spec should not be empty")
+	}
+	// extract PEM from request object
+	block, _ := pem.Decode(csr)
+	if block == nil || block.Type != "CERTIFICATE REQUEST" {
+		return nil, fmt.Errorf("PEM block type must be CERTIFICATE REQUEST")
+	}
+	return x509.ParseCertificateRequest(block.Bytes)
+}
+
+// areAddressesEqual returns true if the dns address in CSR subject matches with the instance address
+// If the address passed is an IP address, we do a reverse lookup to find the host address
+func (c *Approver) areAddressesEqual(instanceAdd string, csrAddr string) bool {
+	// reverse lookup the host if the address is an IP address
+	if parseAddr := net.ParseIP(instanceAdd); parseAddr != nil {
+		dnsAddresses, _ := net.LookupAddr(instanceAdd)
```

Comment:
Why are you ignoring the error? I am also struggling to understand the flow. Won't this fail if the reverse look up fails?

Reply:
I ll add a error check here, I am thinking if it makes more sense to do the platform check here for the fall back option if this fails.